library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/RegistryEntryAttributeReloader.java

```diff
@@ -0,0 +1,229 @@
+
+			AttributeMap attribMap = attributeMaps.computeIfAbsent(attrib,
+					id -> new AttributeMap(registry, attrib));
+			for (var resource : resources) {
+				attribMap.processResource(resource);
+			}
+		}
+	}
+
+	@Override
+	public CompletableFuture<Void> apply(LoadedData data, ResourceManager manager, Profiler profiler, Executor executor) {
+		return CompletableFuture.runAsync(() -> data.apply(profiler), executor);
+	}
+
+	// "<namespace>:attributes/<path>/<file_name>.json" becomes "<namespace>:<file_name>"
+	private Identifier getAttributeId(Identifier jsonId) {
+		String path = jsonId.getPath();
+		int lastSlash = path.lastIndexOf('/');
+		path = path.substring(lastSlash + 1);
+
+		int lastDot = path.lastIndexOf('.');
+		path = path.substring(0, lastDot);
+		return new Identifier(jsonId.getNamespace(), path);
+	}
+
+	protected record LoadedData(Map<RegistryEntryAttribute<?, ?>, AttributeMap> attributeMaps) {
+		public void apply(Profiler profiler) {
+			profiler.push(ID + "/clear_attributes");
+
+			for (var entry : Registry.REGISTRIES.getEntries()) {
+				RegistryEntryAttributeHolderImpl.getData(entry.getValue()).clear();
+			}
+
+			for (Map.Entry<RegistryEntryAttribute<?, ?>, AttributeMap> entry : attributeMaps.entrySet()) {
+				profiler.swap(ID + "/apply_attribute{" + entry.getKey().getId() + "}");
+				applyOne(entry.getKey(), entry.getValue());
+			}
+
+			profiler.pop();
+		}
+
+		@SuppressWarnings("unchecked")
+		private <R, T> void applyOne(RegistryEntryAttribute<R, T> attrib, AttributeMap attribMap) {
+			Registry<R> registry = (Registry<R>) Registry.REGISTRIES.get(attrib.getRegistryKey().getValue());
+			assert registry != null : "huh";
+
+			RegistryEntryAttributeHolderImpl<R> holder = RegistryEntryAttributeHolderImpl.getData(registry);
+			for (Map.Entry<Identifier, Object> attribEntry : attribMap.map.entrySet()) {
+				R item = registry.get(attribEntry.getKey());
+				holder.putValue(item, attrib, (T) attribEntry.getValue());
+			}
+		}
+	}
+
+	protected static final class AttributeMap {
+		private final Registry<?> registry;
+		private final RegistryEntryAttribute<?, ?> attribute;
+		public final Map<Identifier, Object> map;
+
+		public AttributeMap(Registry<?> registry, RegistryEntryAttribute<?, ?> attribute) {
+			this.registry = registry;
+			this.attribute = attribute;
+			map = new HashMap<>();
+		}
+
+		public void processResource(Resource resource) {
+			try {
+				boolean replace;
+				JsonObject values;
+
+				try {
+					JsonObject obj = JsonHelper.deserialize(new InputStreamReader(resource.getInputStream()));
+					replace = JsonHelper.getBoolean(obj, "replace", false);
+					values = JsonHelper.getObject(obj, "values");
+				} catch (JsonSyntaxException e) {
+					LOGGER.error("Invalid JSON file " + resource.getId() + ", ignoring", e);
+					return;
+				}
+
+				if (values == null) {
+					LOGGER.error("Missing 'values' element in {}, ignoring file", resource.getId());
+					return;
+				}
```

Comment:
How would `values` be null here? It's guaranteed by `JsonHelper.getObject` to be not null (since that'll throw if there is no `"values"` key)